cluster/eval_server/add_model.py

```diff
@@ -0,0 +1,155 @@
+    assert os.path.isfile(os.path.join(binary_base, LIB_TF_FW_SO))
+    assert os.path.isfile(os.path.join(binary_base, LIB_TF_CC_SO))
+
+    return True
+
+def copy_to_gcs(src, dst):
+    assert gfile.Exists(src)
+    assert not gfile.Exists(dst)
+
+    with gfile.GFile(src, "rb") as src_f, gfile.GFile(dst, "wb") as dst_f:
+        shutil.copyfileobj(src_f, dst_f)
+
+
+def add_model(name, model_path, flags, binary_path):
+    """Add a player (model + flags) to models cbt
+    name: name of the model
+    model_path: path to model
+    flags: flags used to play this model
+    binary_path: path to gtp binary that can execute model
+    """
+
+    assert verify_params(model_path, flags, binary_path)
+
+    now_date = datetime.datetime.now().isoformat(' ')
+
+    metadata = (
+        (MODEL_NAME, name),
+        (b"model_num", ""),
+        (b"model_flags", flags),
+        (b"run", ""),
+        (b"parent", ""),
+        (b"tag", "eval_server"),
+        (b"tool", "add_model.py"),
+        (b"trained_date", now_date),
+    )
+
+    hash_params = model_path.encode()
+    hash_params += flags.replace(' ', '').replace('=', '').encode()
+    # ten matches MODEL_ROW format
+    player_hash = hashlib.md5(hash_params).hexdigest()[:10]
+    player_name = MODEL_ROW.format("z_eval", player_hash)
+
+    # Check that model doesn't already uploaded.
+    bt_table = (bigtable
+                .Client(FLAGS.cbt_project, admin=True)
+                .instance(FLAGS.cbt_instance)
+                .table(FLAGS.cbt_table))
+    assert bt_table.exists(), f"Table({FLAGS.cbt_table}) doesn't exist"
+    assert bt_table.read_row(player_name) is None, f"{player_name} already uploaded"
+
+    # Upload model and binary
+    gcs_player_folder = os.path.join(PLAYER_FOLDER, player_hash)
+    new_model_path = os.path.join(gcs_player_folder, os.path.basename(model_path))
+    new_binary_base = os.path.join(gcs_player_folder, "bin")
+    binary_base = binary_path.replace(GTP_BIN, "")
+
+    print (f"Saving to {gcs_player_folder!r}")
+
+    #copy_to_gcs(model_path, new_model_path)
+    #for bin_file in [GTP_BIN, LIB_TF_FW_SO, LIB_TF_CC_SO]:
+    #    copy_to_gcs(os.path.join(binary_base, bin_file),
+    #                os.path.join(new_binary_base, bin_file))
+
+    metadata = metadata + (
+        (b"model_path", new_model_path),
+        (b"model_binary_path", new_binary_base),
+    )
+
+    # Add row (READ: player/model) to bigtable
+    row = bt_table.row(player_name)
+    for column, value in metadata:
+        row.set_cell(METADATA, column, value)
+
+    print (metadata)
+    print ()
+    print (row)
+
+    response = bt_table.mutate_rows([row])
+
+    # validate that all rows written successfully
+    for status in enumerate(response):
+        if status.code is not 0:
+            print("Failed to write {}".format(i, status))
+
+
+if __name__ == '__main__':
+    remaining_argv = flags.FLAGS(sys.argv, known_only=True)
+    fire.Fire({
```

Comment:
absl.app?  I'd like to avoid new uses of fire if possible.

Reply:
Let's discuss this offline but now passing args is harder :/

Reply:
Fixed.